src/main/scala/io/citrine/lolo/trees/DecisionPath.scala

```diff
@@ -0,0 +1,136 @@
+  * paths of decisions that depend on those excluded features are assigned non-zero weights that are proportional to
+  * the share of the training data that followed that path.
+  *
+  * This class captures those weights for a single feature and a single node.  The `featureIndex` identifies
+  * which feature is having its effect described.  The `weightWhenIncluded` gives the factor of the weight of the node
+  * due to this feature when the feature is included (i.e. known).  It is always 0 or 1.  The `weightWhenExcluded` gives
+  * the factor of the weight of the node due to this feature when the feature is excluded (i.e. unknown).  It is always
+  * strictly greater than 0 and strictly less than 1.
+  *
+  * @param featureIndex index of the feature this node describes.
+  * @param weightWhenExcluded fraction of paths flowing through this node with this feature excluded.
+  * @param weightWhenIncluded fraction of one paths flowing through this node with this feature included.
+  */
+case class FeatureWeightFactor(
+                          featureIndex: Int,
+                          weightWhenExcluded: Double,
+                          weightWhenIncluded: Double
+                 ) {
+  require(weightWhenIncluded == 0.0 || weightWhenIncluded == 1.0,
+    s"Got weightWhenIncluded=$weightWhenIncluded, but should only ever be 0.0 or 1.0")
+  require(weightWhenExcluded > 0 && weightWhenExcluded < 1,
+    s"Got weightWhenExcluded=$weightWhenExcluded, but should be > 0 and < 1")
+}
+
+/**
+  * Path of unique features used in splitting to arrive at a node in a decision tree
+  *
+  * Note: DecisionPath.size is the number of splitting features. This does NOT include the first
+  * node, which accounts for the "empty path" in computing TreeSHAP.
+  *
+  * The path is extended as new features are encountered, but it actually isn't supposed to represent the sequence
+  * of features that are added.  Rather, each element in `weightBySubsetSize` corresponds to all of the subsets of the
+  * features that are included vs excluded of a given size:
+  *  - the 0th element corresponds to the path where all of the encountered features are excluded,
+  *  - the 1st element sums the paths that have exactly one of the encountered feature included,
+  *  - the 2nd element sums the paths that have exactly two of the encountered features included,
+  * and so on and so forth.  The final element represents the single case where all of the features are "turned on".
+  * This arrangement is convenient for computing combinatorial factors that depend on the feature subset size.
+  *
+  * @param numFeatures number of features in the input space (used for pre-allocation)
+  */
+class DecisionPath(numFeatures: Int) {
+  // pre-allocation of this whole array is an attempted performance optimization.
+  var features: mutable.Set[FeatureWeightFactor] = mutable.Set.empty
```

Comment:
It might be better to have constant-time lookup by index. This comment is not blocking.

Reply:
Yeah; I thought about refactoring it into a map.